README.md
# ReCalc v1.4.0

Calculadora con funciones simples.

## Dependencias

- node > 18

## Comandos

```bash
# iniciar/limpiar la db
npm run seed

# iniciar calculadora
npm run start

# correr lost test
npm run test
```
Cangaro Ignacio Gabriel,
Dominguez Dolores,
Kral Franco,
Libano Santiago Matias.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.4.1] 12-07-2023

### Added

- Test e2e de error en División por 0
- Test e2e de error en Potencia por 200000

## [1.4.0] 17-06-2023

### Added

- Funcionalidades de suma, division, multiplicacion y potencia para Frontend
- Boton de Potencia
- Raiz cuadrada

### Fixed

- Boton de C (clear)
- Errores de guardado en BDD

## [1.3.2] 08-06-2023

### Fixed

- Path donde corre eslint

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.1] 05-06-2023

### Added

- Dependencia cross-env para correr tests e2e

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.0] 05-06-2023

### Added

- Interfaz básica de la calculadora
- ESLint para tests estáticos
- Playwright para tests e2e

## [1.2.0] 15-05-2023

### Added

- Sequelize para el manejo de base de datos
- Tests de integración sobre API y modelo
- Integración continua

## [1.1.0] 20-04-2023

### Added

- Endpoint para la función de resta
- Dependencias para realizar tests
- Tests para la función de resta

## [1.0.0] 17-04-2023

### Added

- Interface CLI para realizar cálculos en forma interactiva
- Base de la API
- Funciones core para realizar suma, resta, multiplicación, división y potencia

[unreleased]: https://github.com/Nevy48/recalc/compare/v1.4.1...HEAD
[1.4.1]:https://github.com/Nevy48/recalc/releases/tag/v1.4.1
[1.4.0]: https://github.com/Nevy48/recalc/releases/tag/v1.4.0
[1.3.2]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.2
[1.3.1]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.1
[1.3.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.3.0
[1.2.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.2.0
[1.1.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.1.0
[1.0.0]: https://github.com/frlp-utn-ingsoft/recalc/releases/tag/v1.0.0


test/ui.test.js
import { test, expect } from '@playwright/test';
import { seed } from '../src/seed.js';
import { Operation, History } from '../src/models.js'

test.describe('test', () => {
  test.describe.configure({ mode: 'serial' });

  test.beforeEach(async () => {
    await seed();
  })

  test('Deberia tener como titulo de pagina recalc', async ({ page }) => {
    await page.goto('./');

    // Expect a title "to contain" a substring.
    await expect(page).toHaveTitle(/recalc/i);
  });

  test('Deberia poder realizar una resta', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '9' }).click()
    await page.getByRole('button', { name: '-' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sub/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(70);

    await expect(page.getByTestId('display')).toHaveValue(/70/)

    const operation = await Operation.findOne({
      where: {
        name: "SUB"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(79)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(70)
  });
  
  test('Deberia poder realizar una suma', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '+' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/add/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(14);

    await expect(page.getByTestId('display')).toHaveValue(/14/)

    const operation = await Operation.findOne({
      where: {
        name: "ADD"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(5)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(14)
  });

  test('Deberia poder realizar una multiplicación', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '*' }).click()
    await page.getByRole('button', { name: '4' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/mul/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(20);

    await expect(page.getByTestId('display')).toHaveValue(/20/)

    const operation = await Operation.findOne({
      where: {
        name: "MUL"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(5)
    expect(historyEntry.secondArg).toEqual(4)
    expect(historyEntry.result).toEqual(20)
  });

  test('Deberia poder realizar una division', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '6' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '2' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/div/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(30);

    await expect(page.getByTestId('display')).toHaveValue(/30/)

    const operation = await Operation.findOne({
      where: {
        name: "DIV"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(60)
    expect(historyEntry.secondArg).toEqual(2)
    expect(historyEntry.result).toEqual(30)
  });

  test('Deberia arrojar error al dividir un numero por 0', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '6' }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '0' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/div/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { error } = await response.json();
    expect(error).toBe(undefined);

    await expect(page.getByTestId('display')).toHaveValue(/undefined/)
  });

  test('Deberia arrojar error al hacer una potencia de un numero mayor a 200000', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '2' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '^' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/pow/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { error } = await response.json();
    expect(error).toBe(undefined);

    await expect(page.getByTestId('display')).toHaveValue(/undefined/)
  });



  test('Debería limpiar el display', async ({ page }) => {
    await page.goto('./');
    await page.getByRole('button', { name: '7' }).click();
    await page.getByRole('button', { name: 'c' }).click();

    await page.waitForSelector('[data-testid="display"]');
    await expect(page.getByTestId('display')).toHaveValue('');
  });

  test('Debería poder realizar una sqrt', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '1' }).click()
    await page.getByRole('button', { name: '6' }).click()
    await page.getByRole('button', { name: '√' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sqrt/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(4);

    await expect(page.getByTestId('display')).toHaveValue(/4/)

    const operation = await Operation.findOne({
      where: {
        name: "SQRT"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(16)
    expect(historyEntry.secondArg).toEqual(null)
    expect(historyEntry.result).toEqual(4)
  });

  test('Debería poder realizar una potencia', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '2' }).click()
    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '^' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/pow/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(625);

    await expect(page.getByTestId('display')).toHaveValue(/625/)

    const operation = await Operation.findOne({
      where: {
        name: "POW"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(25)
    expect(historyEntry.secondArg).toEqual(null)
    expect(historyEntry.result).toEqual(625)
  });


})

test/api.test.js
const request = require('supertest');
const api = require('../src/api.js');
const { seed } = require('../src/seed.js')

beforeEach(async () => {
    await seed()
})

describe("API substract", () => {
    test("Deberia responder con un 200 ok", async () => {
        const app = await api.build()

        return request(app)
            .get('/api/v1/sub/2/1')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(1);
            });
    });
});

describe("API add", () => {
    test("Debería responder con un 200 ok", async () => {
        const app = await api.build();

        return request(app)
            .get('/api/v1/add/2/1')
            .expect(200)
            .expect("Content-Type", "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(3);
            });
    });

    //En el endpoint de suma, probar que si el segundo parámetro es negativo, 
    //el resultado sea menor al primer parámetro y que el endpoint devuelva un status 200.
    test("Debería responder con un 200 ok y el resultado menor al primer parámetro si el segundo parámetro es negativo", async () => {
        const app = await api.build();

        return request(app)
            .get('/api/v1/add/5/-3')
            .expect(200)
            .expect("Content-Type", "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeLessThan(5);
            });
    });

    //Testear que haciendo una request al endpoint de suma con 0.1 y 0.2 el resultado sea 0.3.
    test("Debería responder con un 200 ok y el resultado 0.3 si el primer parámetro es 0.1 y el segundo 0.2", async () => {
        const app = await api.build();

        return request(app)
            .get('/api/v1/add/0.1/0.2')
            .expect(200)
            .expect("Content-Type", "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeCloseTo(0.3, 2);
            });
    });
});

    //En el endpoint de multiplicación, probar que si ambos parámetros tienen decimales,
    //el resultado sea con decimales y que el endpoint devuelva un status 200.
    
    describe("API mul", () => {
        test("Debería responder con un 200 ok", async () => {
            const app = await api.build();
    
            return request(app)
                .get('/api/v1/mul/2/1')
                .expect(200)
                .expect("Content-Type", "application/json; charset=utf-8")
                .then((res) => {
                    expect(res.body.result).toEqual(2);
                });
        });

        function tieneDecimales(num) {
            return num % 1 !== 0;
        }

        test("Debería responder con un 200 ok y si el resultado da decimales, teniendo ambos parámetros con decimales", async () => {
            const app = await api.build();
        
            return request(app)
                .get('/api/v1/mul/1.5/2.5')
                .expect(200)
                .expect("Content-Type", "application/json; charset=utf-8")
                .then((res) => {
                    expect(tieneDecimales(res.body.result)).toBeTruthy();
                });
        });
});

describe("API pow", () => {
    test("Debería responder con un 400 error", async () => {
        const app = await api.build();

        return request(app)
        .get('/api/v1/pow/P')
        .expect(400)
        .expect("Content-Type", "application/json; charset=utf-8")
        .then((res) => {
            expect(res.body.message).toEqual("El parámetro no es un número");
        });
    });
});

describe("API div", () => {
    test("Debería responder con un 400 error", async () => {
      const app = await api.build();
  
      return request(app)
        .get('/api/v1/div/1/0')
        .expect(400)
        .expect("Content-Type", "application/json; charset=utf-8")
        .then((res) => {
          expect(res.body.message).toEqual("Error: No se puede dividir por 0");
        });
    });
  });

describe("API sqrt", () => {
    test("Debería responder con un 200 ok", async () => {
        const app = await api.build();

        return request(app)
        .get('/api/v1/sqrt/4')
        .expect(200)
        .expect("Content-Type", "application/json; charset=utf-8")
        .then((res) => {
            expect(res.body.result).toEqual(2);
        });
    });
});

test/core.test.js
import core from '../src/core.js'

//Test de Substract
describe('Subtract', () => {
    test('Deberia 2 - 2 = 0', () => {
        expect(core.sub(2, 2)).toBe(0); 
    });

    test('Deberia 6 - 4 = 2', () => {
        expect(core.sub(6, 4)).toBe(2); 
    });

	test('Deberia 2 - 4 = Negativo', () => {
		expect(core.sub(2, 4)).toBeLessThan(0);
	});
    test('Deberia lanzar error al intentar usar un parámetro que no es un número', () => {
        expect(() => core.sub('B', 10)).toThrow('Uno de los parámetros no es un número');
    });
});

//Test de Add
describe('Add', () => {
	test('Deberia 2 + 2 = 4', () => {
		expect(core.add(2, 2)).toBe(4);
	});

	test('Deberia 6 + 4 = 10', () => {
		expect(core.add(6, 4)).toBe(10);
	});
    test('Deberia lanzar error al intentar usar un parámetro que no es un número', () => {
        expect(() => core.add('a', 2)).toThrow('Uno de los parámetros no es un número');
    });
});

//Test de Multiply
describe('Multiply', () => {
    test('Deberia 2 * 2 = 4', () => {
        expect(core.mul(2, 2)).toBe(4); 
    });

    test('Deberia 5 * 3 = 15', () => {
        expect(core.mul(5, 3)).toBe(15); 
    });
    test('Deberia 0 * 5 = 0', () => {
        expect(core.mul(0, 5)).toBe(0); 
    });
    test('Deberia -3 * 4 = -12', () => {
        expect(core.mul(-3, 4)).toBe(-12); 
    });
    test('Deberia lanzar error al intentar usar un parámetro que no es un número', () => {
        expect(() => core.mul('a', 2)).toThrow('Uno de los parámetros no es un número');
    });
    test('Deberia -5 * -8 = 40', () => {
        expect(core.mul(-5, -8)).toBe(40); 
});


 //Test de Division
 describe('Division', () => {
    test('Deberia 2 / 2 = 1', () => {
        expect(core.div(2,2)).toBe(1);
    });

    test ('Deberia 6 / 2 = 3', () => {
        expect(core.div(6,2)).toBe(3);
    });
    test('Deberia 10 / 5 = 2', () => {
        expect(core.div(10,5)).toBe(2);
    });
    test('Deberia lanzar un error al dividir por 0', () => {
        expect(() => { core.div(2,0) }).toThrow('No se puede dividir por 0');
    });
    test('Deberia lanzar error al intentar usar un parámetro que no es un número', () => {
        expect(() => core.div('a', 2)).toThrow('Uno de los parámetros no es un número');
    });
});


//Test de Potencia
describe('Pow', () => {
    test('Deberia 2 * 2 = 4', () => {
        expect(core.pow(2)).toBe(4); 
    });

    test('Deberia 5 * 5 = 25', () => {
        expect(core.pow(5)).toBe(25); 
    });

    test('Deberia 7 * 7 = 49', () => {
        expect(core.pow(7)).toBe(49); 
    });

    test('Deberia 10 * 10 = 100', () => {
        expect(core.pow(10)).toBe(100); 
    });

    test('Deberia 4 * 4 = 16', () => {
        expect(core.pow(4)).toBe(16); 
    });

	test('Deberia dar como resultado un numero positivo', () => {
		expect(core.pow(-5)).toBeGreaterThan(0);
	});

    test('Deberia lanzar error al intentar usar un parámetro que no es un número', () => {
        expect(() => core.pow('a')).toThrow('El parámetro no es un número');
    });
})});

//Test de Raiz Cuadrada
describe('Sqrt', () => {
    test('Deberia √16 = 4', () => {
        expect(core.sqrt(16)).toBe(4); 
    });

    test('Deberia √625 = 25', () => {
        expect(core.sqrt(625)).toBe(25); 
    });

    test('Deberia √2401 = 49', () => {
        expect(core.sqrt(2401)).toBe(49); 
    });

    test('Deberia √10000 = 100', () => {
        expect(core.sqrt(10000)).toBe(100); 
    });

    test('Deberia √256 = 16', () => {
        expect(core.sqrt(256)).toBe(16); 
    });

    test('Deberia dar como resultado un numero positivo', () => {
        expect(core.sqrt(25)).toBeGreaterThan(0);
    });

    test('Deberia lanzar error al intentar usar un parámetro que no es un número', () => {
        expect(() => core.sqrt('a')).toThrow('El parámetro no es un número');
    });
});

src/cli.js
import { createInterface } from "node:readline/promises";
import core from "./core.js"

const readline = createInterface({
    input: process.stdin,
    output: process.stdout,
});

const AVAILABLE_FNS = [...Object.keys(core), 'exit'].join(', ')


async function loop() {
    const fnName = await readline.question(`Ingrese función (${AVAILABLE_FNS}): `)

    if (fnName === "exit") {
        console.log("👋👋👋");
        return readline.close();
    }
    if  (fnName !== "add" && fnName !=="sub" && fnName !== "mul" && fnName !== "div" && fnName !== "pow" && fnName !== "sqrt" && fnName !== "exit") {
        console.log("Función erronea");
        return loop();
    }

    const fn = core[fnName];

    const firstNum = await readline.question("Ingrese el primer número: ")
    if (fnName === "pow") {
        const result = fn(Number(firstNum));
        console.log(result);
    }
    else if (fnName === "sqrt") {
        const result = fn(Number(firstNum));
        console.log(result);
    }
    else {
        const secondNum = await readline.question("Ingrese el segundo número: ")
        const result = fn(Number(firstNum), Number(secondNum));
        console.log(result);
    }
    loop();
}

loop();


src/core.js
function add(a, b) {
    if (isNaN(a) || isNaN(b)) {
        throw new Error('Uno de los parámetros no es un número');
    }
	return a + b;
}

function subtract(a, b) {
    if (isNaN(a) || isNaN(b)) {
        throw new Error('Uno de los parámetros no es un número');
    }
    return a - b;
}

function multiply(a, b) {
    if (isNaN(a)) {
        throw new Error("Error: Uno de los parámetros no es un número");
    }
    return a * b;
}

function divide(a, b) {
	if (b === 0) {
		//se verifica que el valor de "b" es 0
		throw new Error("Error: No se puede dividir por 0,");
	}
    if (isNaN(a) || isNaN(b)) {
        throw new Error('Uno de los parámetros no es un número');
    }
	return a / b;
}

function pow(a) {
    if (isNaN(a)) {
      throw new Error("Error: El parámetro no es un número");
    }
  
    if (a > 100000) {
      throw new Error("Error: El número no debe ser mayor a 100.000");
    }
    return a ** 2;
}

function sqrt(a) {
    if (isNaN(a)) {
        throw new Error("Error: El parámetro no es un número");
    }
    return Math.sqrt(a);
}

export default {
    add: add,
    sub: subtract,
    mul: multiply,
    div: divide,
    pow: pow,
    sqrt: sqrt
}


src/models.js
import { Sequelize, DataTypes } from 'sequelize';

const inTest = process.env.NODE_ENV === 'test';

const sequelize = new Sequelize({
    dialect: 'sqlite',
    logging: !inTest,
    storage: inTest ? './db.sqlite3' : './db.sqlite3'
})

export const History = sequelize.define('History', {
    firstArg: {
        type: DataTypes.NUMBER,
        allowNull: false
    },
    secondArg: {
        type: DataTypes.NUMBER,
        allowNull: true
    },
    result: {
        type: DataTypes.NUMBER,
        allowNull: true
    },

    error: { //Agregar un atributo “error” de tipo texto en el modelo History para guardar la de cualquier error que se produzca en una operación
        type: DataTypes.TEXT,
        allowNull: true
    }
});

export const Operation = sequelize.define('Operation', {
    name: {
        type: DataTypes.STRING,
        allowNull: false
    },
});

Operation.hasMany(History)
History.belongsTo(Operation)

export async function createHistoryEntry({ firstArg, secondArg, operationName, result, error}) {
    const operation = await Operation.findOne({
        where: {
            name: operationName
        }
    });

    return History.create({
        firstArg,
        secondArg,
        result,
        OperationId: operation.id,
        error
    })
}

export function createTables() {
    return Promise.all([
        History.sync({ force: true }),
        Operation.sync({ force: true })
    ]);
}

export async function deleteHistory (){
    return History.destroy(
        {   where: {} 
        });
}

export async function getFullHistory() {
    return History.findAll();
}

//Hacer un endpoint para obtener una entrada del historial por id, con el test correspondiente
export async function buscarPorID(historyId) {
    try {
      const history = await History.findByPk(historyId, { include: [Operation] });
  
      if (!history) {
        throw new Error('No se encontró la entrada en el historial');
      }
  
      return history;
    } catch (error) {
      throw new Error('No se pudo recuperar la entrada del historial');
    }
  }
  

src/public/index.js
const $display = document.querySelector('.display')
const $buttons = document.querySelector('.buttons')

const operations = ['-','+','*', '/','√', '^'];

let currentDisplay = "";
let operation = null;
let reset = false;


$buttons.addEventListener('click', async (e) => {
    const nextAction = e.target.name

    if (nextAction === "=") {
        const [firstArg, secondArg] = currentDisplay.split(operation)

        let result;

        if (operation === "-") {
            result = await calculateSub(firstArg, secondArg)
        }

        if (operation === "+") {
            result = await calculateAdd(firstArg, secondArg)
        }

        if (operation === "*") {
            result = await calculateMul(firstArg, secondArg)
        }

        if (operation === "/") {
            result = await calculateDiv(firstArg, secondArg)
        }

        if (operation === "√") {
            result = await calculateSqrt(firstArg)
        }

        if (operation === "^") {
            result = await calculatePow(firstArg)
        }

        reset = true;
        return renderDisplay(result);
    }

    if (nextAction === "c") {
        return clearDisplay();
    }

    if (operations.includes(nextAction)) {
        operation = nextAction;
    }

    if (reset) {
        reset = false;
        operation = null;
        renderDisplay(nextAction);
    } else {
        renderDisplay(currentDisplay + nextAction);
    }
})

async function calculateSub(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/sub/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateDiv(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/div/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateAdd(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/add/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateMul(firstArg, secondArg) {
    const resp = await fetch(`/api/v1/mul/${firstArg}/${secondArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculateSqrt(firstArg) {
    const resp = await fetch(`/api/v1/sqrt/${firstArg}`)
    const { result } = await resp.json();

    return result;
}

async function calculatePow(firstArg) {
    const resp = await fetch(`/api/v1/pow/${firstArg}`);
    const { result } = await resp.json();
  
    return result;
}

function renderDisplay(chars) {
    currentDisplay = chars;
    $display.value = chars;
}

function clearDisplay() {
    currentDisplay = "";
    $display.value = "";
}


src/controllers.js
import express from 'express';
import core from './core.js';

import { createHistoryEntry, deleteHistory, getFullHistory} from './models.js'
import { buscarPorID } from './models.js'

const router = express.Router();

router.get("/sub/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send('Uno de los parámetros no es un número');
    } else {
        const result = core.sub(a, b);

        await createHistoryEntry({ firstArg: a, secondArg: b, result, operationName: "SUB" })
        return res.send({ result });
    }
});


router.get("/add/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send('Uno de los parámetros no es un número');
        await createHistoryEntry({ firstArg: a, secondArg: b, result: null, operationName: "ADD" });
    } else {
        const result = core.add(a, b);
        await createHistoryEntry({ firstArg: a, secondArg: b, result: result, operationName: "ADD" });
        return res.send({ result });
    }


});

router.get("/mul/:a/:b", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        res.status(400).send('Uno de los parámetros no es un número');
        await createHistoryEntry({ firstArg: a, secondArg: b, result: null, operationName: "MUL"});
    } else {
        const result = core.mul(a, b);
        await createHistoryEntry({ firstArg: a, secondArg: b, result: result, operationName: "MUL"});
        return res.send({ result });
    }
});

router.get("/div/:a/:b", async function(req, res) {
    const params = req.params;
    const a = Number(params.a);
    const b = Number(params.b);

    if (isNaN(a) || isNaN(b)) {
        return res.status(400).json({ message: 'Uno de los parámetros no es un número' });
    } 
    if (b === 0) {
        await createHistoryEntry({ firstArg: a, secondArg: b, result: null, operationName: "DIV" });
        return res.status(400).json({ message: 'Error: No se puede dividir por 0' });
    }
    const result = core.div(a, b);
    await createHistoryEntry({ firstArg: a, secondArg: b, result: result, operationName: "DIV" });
    return res.json({ result });
});

router.get("/pow/:a", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);

    if (isNaN(a)) {
        res.status(400).json({message:'El parámetro no es un número'});
    } else if (a > 100000) {
        res.status(400).json({message:'El número no debe ser mayor a 100.000'});
    } else  {
        const result = core.pow(a);
        await createHistoryEntry({ firstArg: a, secondArg: null, result, operationName: "POW"});
        return res.send({ result });
    }
});

router.get("/sqrt/:a", async function (req, res) {
    const params = req.params;
    const a = Number(params.a);

    if (isNaN(a)) {
        res.status(400).json({message:'El parámetro no es un número'});
    } else {
        const result = core.sqrt(a);
        await createHistoryEntry({ firstArg: a, secondArg: null, result, operationName: "SQRT"});
        return res.send({ result });
    }
});

router.get("/history/deleted", async function (req, res) {
    await deleteHistory();
    return res.send({ message: "history is deleted"});
});

router.get("/history", async function (req, res) {
    const history = await getFullHistory();
    return res.send({history});
});

//Hacer un endpoint para obtener una entrada del historial por id, con el test correspondiente
router.get("/history/:id", async function (req, res) {
  const historyId = req.params.id;

  try {
    const history = await buscarPorID(historyId);

    return res.send({ history });
  } catch (error) {
    return res.status(404).send({ error: error.message });
  }
});

export default router;

